### ISSUE TITLE
How to create a Logger subclass without $new?

### ISSUE THREAD

**Issue by gangstR** [s-fleck/lgr#55]
Thank you so much for this wonderful package. As a leader of a bilingual data science team using both R and Python regularly, we've appreciated the opportunity to have really consistent logging between both languages thanks to your R contribution. I've tried several logging packages over the years, and this is perhaps the very best for R.

Due to some recent architectural situations we're navigating, I have a need to add some new methods to your `Logger` class. I'm pretty familiar with `R6`, but I wouldn't say I'm an expert; however, I am a veteran OOP programmer going back to the eighties. I appreciate your design decisions and really like the hierarchical support and the `get_logger()` function since it makes it easy to not worry whether someone else in the team started the log or not :-) The downside of your approach is that I've tried at least a couple of approaches to create a subclass so that I can write the custom log event routines, but none of them work. The lack of support for calling `$new` directly (i.e., per your vignette) may mean that this isn't possible (?).

I fully admit this may be less of an issue and more a lack of my understanding of how to approach the problem with `R6` without a standard init/$new approach. Is there a way, for example, to add a `lg$infoplus()` or `lg$traceplus()` method to a subclass of Logger? In my case, such new methods would be used to wrap `msg` with some internal tags while still respecting the `sprintf()` and object/JSON capabilities of your `...` parm in the base methods of `$info()`, etc. Of course, yes, I could fork your package and add in what I want, but perhaps you've thought of this need and could offer good counsel on how one can best extend your class in more traditional ways rather than my gangstR hackish ways.  Again, I deeply appreciate your contribution to the community. I'm happy to offer my help in accomplishing this idea of custom log methods in your package if needed and if I can help in any way.  Best regards!

**Comment by s-fleck** [comment]
Hey thanks :) the design is copied 1:1 from python logging (including the get_logger semantics). 

You can just subclass logger, and then pass the R6 class object to `get_logger()` as the `class` argument. So this should work:
```
myLogger <- R6::R6Class(
  "LoggerGlue",
  inherit = Logger,
  cloneable = FALSE,  
  ...
)

lg <- get_logger("test/loger", myLogger)
```

`get_logger()` just calls the `$new` method of the R6 class, and also puts the logger in the lgr-internal `loggers` environment. This step is crucial so that the whole hierarchical logging works. You can look the source of `LoggerGlue` and `get_logger_glue()` which is already an example implementation of a `Logger` subclass that overrides the logging methods.

Also you probably don't need to fork the whole package. You could just make your own package `gangstrLgr` that imports lgr and only contains the new logger subclass (cross package inheritance works with R6!) and maybe a custom `get_logger()` function so that you don't always have to call `get_logger("na/me", myLogger)` (again, see  `get_logger_glue()`.

Was this helpful?

**Comment by gangstR** [comment]
Ahhh! __Ich danke Ihnen!__ Your example helped me see the error of my ways.  I guess I was cognitively spent last night after a long week. It seems the only detail I'd missed in one of several attempts was to simply  _name_ the logger with the `name` parameter.  For example, I had tried issuing `get_logger(class = myLogger)` at some point and that alone doesn't work (i.e., I was still seeing a class of `Logger`). I tried several variations and other approaches before switching off my system, but my defined subclass was never showing up and I just didn't see why. When I saw your response, out of curiosity I also tried using `name = "root"` (which won't work either), but then reviewed the source again to see why. That's no big deal in my situation, and I understand why after studying the source with a fresh. well-rested brain. I get it all now. 

I'm glad I didn't pop a cap in my keyboard because writing R code and getting help from the cool R community makes me so happy. Thank you for being both awesome _and_ patient!
